### PR TITLE
Move tensor types check in set_default_tensor_type api from c++ to python side.

### PR DESCRIPTION
**Pull request by Shan19900305**
@ezyang Hi, Yang. I have moved the modifications related to set_default_tensor_type to this PR. What do you think? We can discuss it here. Thank you.

Fixes #ISSUE_NUMBER

### CODE REVIEW

File: torch/__init__.py [hunk 1]
```diff
@@ -745,6 +745,14 @@ def set_default_tensor_type(t):
     """
     if isinstance(t, str):
         t = _import_dotted_name(t)
+    import warnings
+    warnings.warn("torch.set_default_tensor_type() is deprecated as of PyTorch 2.1, "
+                  "please use torch.set_default_dtype() and torch.set_default_device() "
+                  "as alternatives.")
+    # torch.tensortype only alived in c++ side, so using type(torch.FloatTensor)
+    if not isinstance(t, type(torch.FloatTensor)):
```
Comment (ezyang) [inline torch/__init__.py:753]: I'm pretty sure this isn't semantics preserving

**Comment by ezyang** [comment]
I lost the old PR, can you in greater detail motivate this change? (A small sample program showing what this lets you do that you couldn't do before helps.)

**Comment by Shan19900305** [comment]
> I lost the old PR, can you in greater detail motivate this change? (A small sample program showing what this lets you do that you couldn't do before helps.)

Old PR is: https://github.com/pytorch/pytorch/pull/118852, and i want set_default_tensor_type API more general for other device.

File: torch/__init__.py (hunk 1 above)
Comment (Shan19900305) [reply torch/__init__.py:753]: Yep, This check will be more generic for all tensor types. 
The content of the last discussion：
from Shan19900305：
PyTensorType_Check only checks object whether is in tensor_types, and it is assumed that the data stored on the torch._tensor_classes and tensor_types vector are consistent.
Now tensor_types are only used for CPU and GPU tensor types, which is a local static vector. But _tensor_classes is python side, which can be registered from other libraries for custom device tensor types. So from checked object based on local tensor_types vector to global torch._tensor_classes.
from ezyang：
Drop this bit from this PR, if you really want it let's litigate it on its own. But this is the same "I don't want to aid and abet more usage of the legacy constructor style"

And now split this bit to this PR.

**Comment by Shan19900305** [comment]
> In the old PR, I told y'all not to use the legacy Python API. I still stand by this. So I am not convinced this should go in either. Why is this implementation better?

Yes, last time you said that, if you really want it let's litigate it on its own. So i split legacy Python API modification to this pr.

Why is this implementation better?
I think this is more common way to check tensor types, without use loops to check if the type meets the requirements. And more explicitly inform users of this limitation in python code.

**Comment by ezyang** [comment]
I am not convinced. Maybe you can describe your concrete use case and I can tell you other ways to do it.

**Comment by Shan19900305** [comment]
> I am not convinced. Maybe you can describe your concrete use case and I can tell you other ways to do it.

A lot user training script is like this: https://github.com/OpenGVLab/LLaMA-Adapter/blob/8c50ee5d5d393c9bee5fcfda6aaea31d3ca3c40c/gorilla/inference/example.py#L57

And we want using torch.set_default_tensor_type(torch.xxx.HalfTensor) to instead torch.set_default_tensor_type(torch.cuda.HalfTensor).

**Comment by ezyang** [comment]
Use torch.set_default_dtype and the torch.device context manager instead

**Comment by Shan19900305** [comment]
> Use torch.set_default_dtype and the torch.device context manager instead

As your issue: https://github.com/pytorch/pytorch/issues/92701. If we do this, the performance of network maybe decrease. So keep this using this API after community delete this API?

**Comment by ezyang** [comment]
I think I might be amenable to some PRs helping optimize the context manager, it depends on the approach though

**Comment by Shan19900305** [comment]
> I think I might be amenable to some PRs helping optimize the context manager, it depends on the approach though

ok, but still fused about this api. Now this api is widely used, why do we need to discard it?

**Comment by ezyang** [comment]
Default tensor type API does too many things: it simultaneously modifies the dtype AND the device. Also, it makes some policy decisions (what actually gets overridden this way) which cannot be easily tweaked.

**Pull request closed without merging**